### PR TITLE
chore(candles-chart): fill up missing candles

### LINKED ISSUE #3440: Fill missing candles using last close price
## The Chore

There can be gaps in candle data due to either auctions or network upgrades. This makes the chart look broken. We can fill in any missing candle objects with the last known close price.

## Tasks

- [ ] Add candle objects with 0 volume for missing intervals

## Additional details / background info

https://github.com/vegaprotocol/frontend-monorepo/issues/3319#issuecomment-1507147787

### PR DESCRIPTION
**Pull request by macqbat**
# Related issues 🔗

Closes #3440 

# Description ℹ️

Detect discontinuity of data and fill up gaps using last known price

# Demo 📺

-

# Technical 👨‍🔧

-

### CODE REVIEW

File: libs/candles-chart/src/lib/data-source.ts [hunk 1]
```diff
@@ -213,6 +220,95 @@ export class VegaDataSource implements DataSource {
...
+        previousStartDate,
+        candleStartDate
+      );
+      if (difference > 1) {
+        for (let j = 1; j < difference; j++) {
+          const duration = getDuration(interval, j);
+          const newStartDate = add(previousStartDate, duration);
+          agg.push({
+            periodStart: newStartDate.toISOString(),
+            lastUpdateInPeriod: add(newStartDate, duration).toISOString(),
```
Comment (asiaznik) [inline libs/candles-chart/src/lib/data-source.ts:298]: ☝🏻  that's not needed

File: libs/candles-chart/src/lib/data-source.ts [hunk 2]
```diff
@@ -213,6 +220,95 @@ export class VegaDataSource implements DataSource {
...
+        interval,
+        previousStartDate,
+        candleStartDate
+      );
+      if (difference > 1) {
+        for (let j = 1; j < difference; j++) {
+          const duration = getDuration(interval, j);
+          const newStartDate = add(previousStartDate, duration);
+          agg.push({
+            periodStart: newStartDate.toISOString(),
```
Comment (asiaznik) [inline libs/candles-chart/src/lib/data-source.ts:297]: ☝🏻 string -> Date -> string -> Date
Avoidable after `parseCandle` refactor

**Review by bglownia — APPROVED** [review]

File: libs/candles-chart/src/lib/data-source.ts (hunk 2 above)
Comment (macqbat) [reply libs/candles-chart/src/lib/data-source.ts]: has been optimized

File: libs/candles-chart/src/lib/data-source.ts (hunk 1 above)
Comment (macqbat) [reply libs/candles-chart/src/lib/data-source.ts]: has been optimised in some different, much better way

**Pull request merged**